### PR TITLE
Introduce ResolutionOutputs, ResolutionAccess

### PR DESCRIPTION
**Pull request by jvandort**
`DefaultConfiguration` currently implements the 'tail-end' of resolution -- meaning it implements the process of converting the outputs of resolution into user-consumable types like `FileCollection`, `ArtifactCollection`, `ResolvedComponentResult`, etc.

We would like to reuse this implementation elsewhere to provide alternative resolution APIs.

This commit extracts a `ResolutionOutputs` interface which exposes the results of resolution. The intention is for this interface to eventually become public in some form.

We also introduce the concept of a `ResolutionHandle`, which tracks the raw inputs and outputs of resolution. A `ResolutionHandle` is a lazy reference to some resolution and is used as the source of all data for a `ResolutionOutputs`. Eventually, we should be able to construct a `ResolutionHandle` from things other than a `Configuration`.

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

**Comment by jvandort** [comment]
@bot-gradle test this

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/resolver/DefaultResolutionOutputs.java [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+
+        @Override
+        @Deprecated
+        public ImmutableAttributes getAttributes() {
+            DeprecationLogger.deprecateMethod(ArtifactView.class, "getAttributes()")
+                .willBecomeAnErrorInGradle9()
+                .withUpgradeGuideSection(8, "deprecate_artifact_view_get_attributes")
+                .nagUser();
+
+            return computeViewAttributes(resultProvider.getValue());
```
Comment (jvandort) [inline platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/resolver/DefaultResolutionOutputs.java:254]: Its weird that calling this method causes resolution. 
Can we temporarily wire these input attributes to this class until the deprecation happens?

**Comment by jvandort** [comment]
This exposes the same types you can get from a configuration, so files, artifacts, and the root variant of a resolution result. 

This is part of an effort to eventually introduce a new resolver api independent of configurations. 

This pr extracts the logic required to support these resolution outputs into a separate class from configuration itself

**Comment by jvandort** [comment]
One downside to resolution result is that the whole type is not lazy or cc compatible. Only getRootComponent is. 

This type exposes only CC compatible and lazy results

**Comment by jvandort** [comment]
In practice ResolutionResult is and will continue to be relevant. RR provides access to the resolved graph via the root component. Future APIs will provide access to the graph via the same component/variant APIs.

**Comment by jvandort** [comment]
There is no timeline to when this new API will be introduced. Until then RR will remain fully supported. We expect the transition to remain simple, as both APIs expose the same data

**Review by lkasso — APPROVED** [review]
Docs LGTM.

**Comment by jvandort** [comment]
@bot-gradle test this

**Comment by jvandort** [comment]
@bot-gradle test this

**Comment by jvandort** [comment]
@bot-gradle test this

**Comment by jvandort** [comment]
@bot-gradle test this

**Review by ghale — APPROVED** [review]
LGTM

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/resolver/ResolutionHandle.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+import org.gradle.api.internal.artifacts.ResolverResults;
+import org.gradle.api.internal.artifacts.configurations.ResolutionHost;
+import org.gradle.api.internal.artifacts.configurations.ResolutionResultProvider;
+import org.gradle.api.internal.attributes.ImmutableAttributes;
+
+/**
+ * An internal lazy reference to a graph resolution. Provides access to the inputs and
+ * outputs of a graph resolution.
+ */
+public interface ResolutionHandle {
```
[the remark this answers is not in the document]
Comment (jvandort) [reply platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/resolver/ResolutionHandle.java]: Awh dang I kinda thought the name was cool, but `ResolutionAccess` is just as good. Changed the name

**Pull request merged**